File: python/modcdp/injector/DiscoverExtensionInjector.py
# MODCDP_TRANSLATE: KEEP THIS FILE TRANSLATED ACROSS TYPESCRIPT, PYTHON, AND GO.
# Keep all shapes, signatures, behavior, and tests 1:1 in sync with:
# - ./js/src/injector/DiscoverExtensionInjector.ts
# - ./go/modcdp/injector/DiscoverExtensionInjector.go
from __future__ import annotations

from ..injector.ExtensionInjector import (
    ExtensionInjector,
    ExtensionInjectionResult,
    InjectorConfig,
)
from ..injector.NodeExtensionFiles import (
    PreparedExtension,
    extensionIdFromManifestKey,
    prepareUnpackedExtension,
)


class DiscoverExtensionInjector(ExtensionInjector):
    def __init__(self, config: InjectorConfig | dict | None = None) -> None:
        config = config.model_dump() if isinstance(config, InjectorConfig) else dict(config or {})
        super().__init__({**config, "injector_mode": "discover"})
        self.prepared_extension: PreparedExtension | None = None

    def prepare(self) -> None:
        extension_path = self.config.injector_discover_extension_path
        if not self.config.injector_service_worker_extension_id and extension_path:
            manifest_path = extension_path
            if extension_path.endswith(".zip"):
                self.prepared_extension = prepareUnpackedExtension(extension_path)
                manifest_path = self.prepared_extension.unpacked_extension_path
            self.service_worker_extension_id = extensionIdFromManifestKey(manifest_path)
        super().prepare()

    def inject(self) -> ExtensionInjectionResult | None:
        discovered = self._discoverReadyServiceWorker()
        if discovered:
            discovered.source = "discover"
            return discovered
        if self.config.injector_trust_service_worker_target:
            waited = self._waitForReadyServiceWorker(
                self.config.injector_service_worker_probe_timeout_ms,
                matched_only=True,
            )
            if waited:
                waited.source = "discover"
                return waited
        if not self.config.injector_require_service_worker_target:
            return None
        waited = self._waitForReadyServiceWorker(
            self.config.injector_service_worker_ready_timeout_ms,
            matched_only=self.config.injector_trust_service_worker_target,
        )
        if waited:
            waited.source = "discover"
            return waited
        matchers = ", ".join(
            [
                *self.config.injector_service_worker_url_includes,
                *self.config.injector_service_worker_url_suffixes,
            ]
        )
        raise RuntimeError(f"Required ModCDP service worker target was not visible ({matchers or 'no matcher'}).")

    def close(self) -> None:
        super().close()
        if self.prepared_extension:
            self.prepared_extension.cleanup()
            self.prepared_extension = None

File: go/modcdp/injector/DiscoverExtensionInjector.go
// MODCDP_TRANSLATE: KEEP THIS FILE TRANSLATED ACROSS TYPESCRIPT, PYTHON, AND GO.
// Keep all shapes, signatures, behavior, and tests 1:1 in sync with:
// - ./js/src/injector/DiscoverExtensionInjector.ts
// - ./python/modcdp/injector/DiscoverExtensionInjector.py
package injector

import (
	"fmt"
	"os"
	"strings"
)

type DiscoverExtensionInjector struct {
	ExtensionInjector
	CleanupPath string
}

func NewDiscoverExtensionInjector(config InjectorConfig) DiscoverExtensionInjector {
	config.InjectorMode = "discover"
	return DiscoverExtensionInjector{ExtensionInjector: NewExtensionInjector(config)}
}

func (i *DiscoverExtensionInjector) Prepare() error {
	extensionPath := i.Config.InjectorDiscoverExtensionPath
	if i.Config.InjectorServiceWorkerExtensionID == "" && extensionPath != "" {
		manifestPath := extensionPath
		if strings.HasSuffix(extensionPath, ".zip") {
			prepared, err := PrepareUnpackedExtension(extensionPath)
			if err != nil {
				return err
			}
			manifestPath = prepared.UnpackedExtensionPath
			i.CleanupPath = prepared.CleanupPath
		}
		extensionID, err := ExtensionIDFromManifestKey(manifestPath)
		if err != nil {
			return err
		}
		i.ServiceWorkerExtensionID = extensionID
	}
	return nil
}

func (i *DiscoverExtensionInjector) Inject() (*ExtensionInjectionResult, error) {
	discovered, err := i.discoverReadyServiceWorker(false)
	if err != nil || discovered != nil {
		if discovered != nil {
			discovered.Source = "discover"
		}
		return discovered, err
	}
	if i.Config.InjectorTrustServiceWorkerTarget {
		waited, err := i.waitForReadyServiceWorker(i.Config.InjectorServiceWorkerProbeTimeoutMS, true)
		if err != nil || waited != nil {
			if waited != nil {
				waited.Source = "discover"
			}
			return waited, err
		}
	}
	if !i.Config.InjectorRequireServiceWorkerTarget {
		return nil, nil
	}
	waited, err := i.waitForReadyServiceWorker(i.Config.InjectorServiceWorkerReadyTimeoutMS, i.Config.InjectorTrustServiceWorkerTarget)
	if err != nil || waited != nil {
		if waited != nil {
			waited.Source = "discover"
		}
		return waited, err
	}
	matchers := append(append([]string{}, i.Config.InjectorServiceWorkerURLIncludes...), i.Config.InjectorServiceWorkerURLSuffixes...)
	matcherText := strings.Join(matchers, ", ")
	if matcherText == "" {
		matcherText = "no matcher"
	}
	return nil, fmt.Errorf("required ModCDP service worker target was not visible (%s)", matcherText)
}

func (i *DiscoverExtensionInjector) Close() error {
	if i.CleanupPath != "" {
		_ = os.RemoveAll(i.CleanupPath)
		i.CleanupPath = ""
	}
	return nil
}
